### ISSUE TITLE
unusual_byte_groupings triggers for binary literals

### ISSUE THREAD

**Issue by sdroege** [rust-lang/rust-clippy#11749]
### Summary

This is kind of related to https://github.com/rust-lang/rust-clippy/issues/9183 and https://github.com/rust-lang/rust-clippy/issues/11243.

When I'm using binary literals for parsing formats, I usually don't use equal groupings but instead use groupings that reflect the underlying format.

It's easy enough to allow `unusual_byte_groupings` for these parts of the code, but I'm probably not the only one doing that and maybe the lint should be disabled completely for binary literals or only trigger in more restricted cases (all groups the same but one is off-by-one or so?).

### Lint Name

unusual_byte_groupings

### Reproducer

The following example is from parsing the [header](http://www.mp3-tech.org/programmer/frame_header.html) of an MP3 frame, specifically parsing the layer description bits.

The second byte of the header contains 4 groups: 3 bits, 2 bits, 2 bits (what we're interested in here) and another bit in the end so I'd use `0b000_00_11_0` as a mask.

```rust
fn main() {
    let data = [0u8; 4];
    
    let layer_desc = (data[1] & 0b000_00_11_0) >> 1;
    println!("meh {layer_desc}");
}
```

I saw this happen:

```
warning: digits of hex, binary or octal literal not in groups of equal size
 --> src/main.rs:4:33
  |
4 |     let layer_desc = (data[1] & 0b000_00_11_0) >> 1;
  |                                 ^^^^^^^^^^^^^ help: consider: `0b0000_0110`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unusual_byte_groupings
  = note: `#[warn(clippy::unusual_byte_groupings)]` on by default
```

I expected to see this happen: [nothing]

### Version

```text
rustc 1.73.0 (cc66ad468 2023-10-03)
binary: rustc
commit-hash: cc66ad468955717ab92600c770da8c1601a4ff33
commit-date: 2023-10-03
host: x86_64-unknown-linux-gnu
release: 1.73.0
LLVM version: 17.0.2
```

**Comment by blyxyas** [comment]
To quote what I said in #12175:

> Honestly, I think that the best way to proceed here is to close the issue and to accept that Clippy cannot comprehend intent at this level. We could keep adding edge-cases, but there are so many arbitrary patterns in binary data encoding that there's no way of knowing.

I'll close this, this is a more than valid scenario to use `#[allow(unusual_byte_groupings)]`, the `-- -Aunusual_byte_groupings` or use the [`[lints]`](https://doc.rust-lang.org/nightly/cargo/reference/manifest.html#the-lints-section) table in `Cargo.toml`

Thanks for reporting the issue nonetheless ( ^..^)ﾉ

**Comment by sdroege** [comment]
I think that makes sense, however I would've rather disabled the lint by default for binary literals instead :)

**Comment by rcgoodfellow** [comment]
If Clippy cannot comprehend intent at this level, this lint should not be enabled by default. There are no edge cases, any grouping is valid. In my experience, even groupings are the exception and not the rule.

**Comment by Omnikron13** [comment]
Just found this when searching before preparing to raise the exact same issue...

The fact is the matter is that with _binary literals specifically_, the intent of the programmer is basically _never_ to express a simple number.

To be totally clear: **_binary literals are for representing bit patterns._**

Bit patterns are not mentally scanned in the same way as large numbers, where e.g. a 'thousand separator' helps to grok the order of magnitude, and they are most often only covering a byte at a time, where the semantics of 'unusual' separation of the 8 bits is far more important than simple visual separation.

This lint is badly misapplied for binary literals, imo.

(A more appropriate lint for them would be warning about having literals that are more of less than 8 bits)

**Comment by nk9** [comment]
Are you still working on this, @nshah067? If not, I can take a look.
